lib/SIL/Utils/OwnershipUtils.cpp

```diff
@@ -1994,6 +2016,355 @@ bool swift::visitAdjacentReborrowsOfPhi(
+                                          ValueSetVector &visitedValues) {
+    // Phi cycles are skipped. They cannot contribute any new enclosing defs.
+    if (!visitedPhis.insert(forwardingPhi))
+      return false;
+
+    bool foundIntroducer = false;
+    SILBasicBlock *block = forwardingPhi->getParent();
+    for (auto *pred : block->getPredecessorBlocks()) {
+      SILValue incomingValue = forwardingPhi->getIncomingPhiValue(pred);
+
+      // Each phi operand requires a new introducer list and visited values
+      // set. These values will be remapped to successor phis before adding them
+      // to the caller's introducer list. It may be necessary to revisit a value
+      // that was already visited by the caller before remapping to phis.
+      StackList<SILValue> incomingIntroducers(function);
+      ValueSetVector incomingVisitedValues;
+      if (!recursivelyFindBorrowIntroducers(incomingValue, incomingIntroducers,
+                                            incomingVisitedValues))
+        continue;
+
+      foundIntroducer = true;
+      findSuccessorDefsFromPredDefs(pred, incomingIntroducers, introducers,
+                                    visitedValues);
+    }
+    return foundIntroducer;
+  }
+
+  // Given a reborrow operand's incoming value, find the enclosing definition.
+  void recursivelyFindDefsOfReborrowOperand(
+    SILValue incomingValue,
+    StackList<SILValue> &enclosingDefs) {
+
+    if (incomingValue->getOwnershipKind() == OwnershipKind::None)
+      return;
+
+    assert(incomingValue->getOwnershipKind() == OwnershipKind::Guaranteed);
+
+    // Avoid repeatedly constructing BorrowedValue during use-def
+    // traversal. That would be quadratic if it checks all uses for reborrows.
+    if (auto *predPhi = dyn_cast<SILPhiArgument>(incomingValue)) {
+      recursivelyFindDefsOfReborrow(predPhi, enclosingDefs);
+      return;
+    }
+
+    // Handle non-phi borrow introducers.
+    BorrowedValue borrowedValue(incomingValue);
+
+    switch (borrowedValue.kind) {
+    case BorrowedValueKind::Phi:
+      llvm_unreachable("phis are short-curcuited above");
+    case BorrowedValueKind::Invalid:
+      llvm_unreachable("A reborrow immediate operand must be a BorrowedValue.");
+    case BorrowedValueKind::BeginBorrow: {
+      ValueSetVector visitedValues;
+      recursivelyFindBorrowIntroducers(
+        cast<BeginBorrowInst>(incomingValue)->getOperand(), enclosingDefs,
+        visitedValues);
+      break;
+    }
+    case BorrowedValueKind::LoadBorrow:
+    case BorrowedValueKind::SILFunctionArgument:
+      // There is no enclosing def on this path.
+      break;
+    }
+  }
+
+  // Given a reborrow, find the definitions of the enclosing borrow scopes. Each
+  // enclosing borrow scopes is represented by either a dominating owned value,
+  // an owned phi in the same block, a dominating outer borrow introducer, or an
+  // outer-adjacent reborrow.
```

Comment:
Might be informative to associate one of the example registers below with each case there's an example for:
- dominating owned value -> no example
- owned phi in the same block -> `%phi_3` (is this an "owned adjacent phi"?)
- dominating outer borrow introducer -> `%outerBorrowB`
- outer-adjacent reborrow -> `%outerReborrow`

Reply:
Rewritten:
```
  // Given a reborrow, find the definitions of the enclosing borrow scopes. Each
  // enclosing borrow scope is represented by one of the following cases, which
  // refer to the example below:
  //
  // dominating owned value -> %value encloses %reborrow_1
  // owned outer-adjacent phi -> %phi_3 encloses %reborrow_3
  // dominating outer borrow introducer -> %outerBorrowB encloses %reborrow
  // outer-adjacent reborrow -> %outerReborrow encloses %reborrow
  //
  // Recurse along the use-def phi web until a begin_borrow is reached. Then
  // find all introducers of the begin_borrow's operand. At each level, find
  // the outer adjacent phi, if one exists, otherwise return the most recently
  // found dominating definition.
  //
  // If \p reborrow was already encountered because of a phi cycle, then no
  // enclosingDefs are added.
  //
  // Example:
  //
  //         %value = ...
  //         %borrow = begin_borrow %value
  //         br one(%borrow)
  //     one(%reborrow_1 : @guaranteed)
  //         br two(%value, %reborrow_1)
  //     two(%phi_2 : @owned, %reborrow_2 : @guaranteed)
  //         br three(%value, %reborrow_1)
  //     three(%phi_3 : @owned, %reborrow_3 : @guaranteed)
  //         end_borrow %reborrow_3
  //         destroy_value %phi_3
  //
  // recursivelyFindDefsOfReborrow(%reborrow_3) returns %phi_3 by
  // computing enclosing defs (inner -> outer) in this order:
  //
  //     %reborrow_1 -> %value
  //     %reborrow_2 -> %phi_2
  //     %reborrow_3 -> %phi_3
  //
  // Example:
  //
  //         %outerBorrowA = begin_borrow
  //         %outerBorrowB = begin_borrow
  //         %struct = struct (%outerBorrowA, outerBorrowB)
  //         %borrow = begin_borrow %struct
  //         br one(%outerBorrowA, %borrow)
  //     one(%outerReborrow : @guaranteed, %reborrow : @guaranteed)
  //
  // recursivelyFindDefsOfReborrow(%reborrow) returns
  // (%outerReborrow, %outerBorrowB).
  //
```